crates/server/src/api/v1/monitor.rs

```diff
@@ -0,0 +1,124 @@
+            .with_state(self)
+    }
+}
+
+struct MonitorApiError(MonitorError);
+
+impl From<DataStoreError> for MonitorApiError {
+    fn from(e: DataStoreError) -> Self {
+        tracing::error!("unexpected data store error: {e}");
+
+        Self(MonitorError::Message {
+            status: StatusCode::INTERNAL_SERVER_ERROR.as_u16(),
+            message: "an error occurred while processing the request".into(),
+        })
+    }
+}
+
+impl IntoResponse for MonitorApiError {
+    fn into_response(self) -> axum::response::Response {
+        (StatusCode::from_u16(self.0.status()).unwrap(), Json(self.0)).into_response()
+    }
+}
+
+#[debug_handler]
+async fn verify_checkpoint(
+    State(config): State<Config>,
+    RegistryHeader(_registry_header): RegistryHeader,
+    Json(body): Json<SerdeEnvelope<TimestampedCheckpoint>>,
+) -> Result<Json<CheckpointVerificationResponse>, MonitorApiError> {
+    // check checkpoint:
+    // - if `log_length` not found, `checkpoint` is `Invalid`;
+    // - if `log_root` or `map_root` is incorrect, `checkpoint` is `Invalid`;
+    // - if `key_id` and `signature` does not match previously stored value,
+    //   `signature` is `Unverified` and will check against the operator log;
+    let (checkpoint_verification, mut signature_verification) = match config
+        .core_service
+        .store()
+        .get_checkpoint(body.as_ref().checkpoint.log_length)
+        .await
+    {
+        Ok(checkpoint) => {
+            // check log root and map root
+            let checkpoint_verification = if checkpoint.as_ref().checkpoint.log_root
+                == body.as_ref().checkpoint.log_root
+                && checkpoint.as_ref().checkpoint.map_root == body.as_ref().checkpoint.map_root
+            {
+                CheckpointVerificationState::Verified
+            } else {
+                CheckpointVerificationState::Invalid
+            };
+
+            // check for exact match on signature and key ID
+            let signature_verification = if checkpoint.signature() == body.signature()
+                && checkpoint.key_id() == body.key_id()
+            {
+                CheckpointVerificationState::Verified
+            } else {
+                // set to Unverified and check against operator log keys below
+                CheckpointVerificationState::Unverified
+            };
+
+            (checkpoint_verification, signature_verification)
+        }
+        Err(DataStoreError::CheckpointNotFound(_)) => (
+            CheckpointVerificationState::Invalid,
+            CheckpointVerificationState::Unverified,
+        ),
+        Err(error) => return Err(MonitorApiError::from(error)),
+    };
+
+    // if `Unverified`, check signature against keys in operator log:
+    //
+    // - if `key_id` is not known or it does not have permission to sign checkpoints or
+    //   the `signature` is invalid, `signature` is `Invalid`;
+    if signature_verification == CheckpointVerificationState::Unverified {
```

Comment:
I'm not sure that this does what  you think it does, though I might be missing something.
Do you have any tests for this logic?

Specifically, I don't see when the `signature_verification` mut variable on line 60 is ever actually used.

Reply:
See lines 77-85 above in the file.

This logic is particular to the design of the reference implementation as it stands today. Every `TimestampedCheckpoint` that was issued is stored. We do a lookup and see if the requested checkpoint is an exact match first. That's where this `signature_verification` could be `Verified` for an exact match or `Unverified` if it is not an exact match. If it is not exact match, we proceed to see if the signature could be valid given what is in the operator log.